File: Collections/Rectangle.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace Collections
{
    /// <summary>
    /// Represents an immutable hyperrectagle. 
    /// </summary>
    public class Rectangle : IEquatable<Rectangle>
    {
        private readonly (double left, double right)[] limits;
        private static readonly EqualityComparer<ValueTuple<double, double>> comparer
            = EqualityComparer<ValueTuple<double, double>>.Default;


        public Rectangle(params (double left, double right)[] limits)
        {
            this.limits = limits;
        }

        public IEnumerable<(double left, double right)> Limits() => limits.Select(i => i);

        public int Dimension() => limits.Length;

        public static Rectangle InfiniteBox(int dimension = 3)
        {
            var limits = Enumerable
                .Range(0, dimension)
                .Select(_ => (double.MinValue, double.MaxValue))
                .ToArray();
            return new Rectangle(limits);
        }

        /// <summary>
        /// Returns a new rectangle with the left boundary of dimension j set to the axis defined by j = v.
        /// </summary>
        public Rectangle SetLeft(int j, double v)
        {
            var truncated = limits
                .Select((t, i) => i == j ? (v, t.right) : t)
                .ToArray();
            return new Rectangle(truncated);
        }

        /// <summary>
        /// Returns a new rectangle with the right boundary of dimension j set to the axis defined by j = v.
        /// </summary>
        public Rectangle SetRight(int j, double v)
        {
            var truncated = limits
                .Select((t, i) => i == j ? (t.left, v) : t)
                .ToArray();
            return new Rectangle(truncated);
        }

        public bool Contains(Point p) => limits
            .Zip(p.Coordinates(), (l, r) => (l, r))
            .All(t => t.l.left <= t.r && t.r <= t.l.right);

        public bool Contains(Rectangle that) => limits
            .Zip(that.limits, (l, r) => (l, r))
            .All(t => t.r.left >= t.l.left && t.r.right <= t.l.right);

        public bool Intersects(Rectangle that) => limits
            .Zip(that.limits, (l, r) => (l, r))
            .All(t => t.r.left <= t.l.right && t.r.right >= t.l.left);

        /// <summary>
        /// Returns the shortest distance squared between this rectangle and point p.
        /// </summary>
        public double DistanceSquaredTo(Point p)
        {
            // The shortest distance is:
            // i) the distance to the closest vertex or edge, if the point is outside the rectangle
            // ii) 0 otherwise
            double Di((double left, double right) bound, double v)
            {
                if (v < bound.left)
                {
                    return bound.left - v;
                }
                else if (v > bound.right)
                {
                    return v - bound.right;
                }
                else
                {
                    return 0;
                }
            }

            return limits
                .Zip(p.Coordinates(), Di)
                .Select(di => Math.Pow(di, 2))
                .Sum();
        }

        public double DistanceTo(Point p) => Math.Sqrt(DistanceSquaredTo(p));

        public override bool Equals(object obj) => this.Equals(obj as Rectangle);

        public bool Equals(Rectangle r)
        {
            // If parameter is null, return false.
            if (Object.ReferenceEquals(r, null))
                return false;

            // Optimization for a common success case.
            if (Object.ReferenceEquals(this, r))
                return true;

            // If run-time types are not exactly the same, return false.
            if (this.GetType() != r.GetType())
                return false;

            return limits.SequenceEqual(r.limits);
        }

        public static bool operator ==(Rectangle lhs, Rectangle rhs)
        {
            // Check for null on left side.
            if (Object.ReferenceEquals(lhs, null))
            {
                if (Object.ReferenceEquals(rhs, null))
                {
                    // null == null = true.
                    return true;
                }

                // Only the left side is null.
                return false;
            }
            // Equals handles case of null on right side.
            return lhs.Equals(rhs);
        }

        public static bool operator !=(Rectangle lhs, Rectangle rhs) => !(lhs == rhs);

        public override int GetHashCode()
        {
            unchecked
            {
                return limits.Aggregate(17, (hash, pair) => hash * 31 + comparer.GetHashCode(pair));
            }
        }
    }
}

File: src/main/java/in/bfdes/collections/Rectangle.java
package in.bfdes.collections;

/**
 * Rectangle abstraction for use in the <code>KdTree</code> class.
 * @param left Lowest coordinate on the x-axis
 * @param right Highest coordinate on the x-axis
 * @param up Highest coordinate on the y-axis
 * @param down Highest coordinate on the y-axis
 */
public record Rectangle(double left, double right, double up, double down) {
    public static final Rectangle INFINITE = new Rectangle(Double.MIN_VALUE, Double.MAX_VALUE, Double.MAX_VALUE, Double.MIN_VALUE);

    public Rectangle withLeft(double left) {
        return new Rectangle(left, right, up, down);
    }

    public Rectangle withRight(double right) {
        return new Rectangle(left, right, up, down);
    }

    public Rectangle withUp(double up) {
        return new Rectangle(left, right, up, down);
    }

    public Rectangle withDown(double down) {
        return new Rectangle(left, right, up, down);
    }

    public boolean contains(Point point) {
        if (point == null)
            throw new IllegalArgumentException();

        if (point.x() < left || point.x() > right)
            return false;
        if (point.y() < down || point.y() > up)
            return false;
        return true;
    }

    public boolean contains(Rectangle rectangle) {
        if (rectangle == null)
            throw new IllegalArgumentException();

        if (rectangle.left() > left || right < rectangle.right())
            return false;
        if (rectangle.up() > up || down > rectangle.down())
            return false;
        return true;
    }

    public boolean intersects(Rectangle rectangle) {
        if (rectangle == null)
            throw new IllegalArgumentException();

        if (rectangle.right() < left || right < rectangle.left())
            return false;
        if (rectangle.up() < down || up < rectangle.down())
            return false;
        return true;
    }

    public double distanceSquaredTo(Point point) {
        if (point == null)
            throw new IllegalArgumentException();

        // The shortest distance is
        // 1) the distance to the closest vertex or edge, if the point is outside the rectangle, or
        // 2) 0 otherwise.
        var dx = 0.0;
        if (point.x() < left)
            dx = left - point.x();
        else if (right < point.x())
            dx = point.x() - right;

        var dy = 0.0;
        if (point.y() < down)
            dy = down - point.y();
        else if (up < point.y())
            dy = point.y() - up;

        return Math.pow(dx, 2) + Math.pow(dy, 2);
    }

    /**
     * Returns the shortest distance between this rectangle and <code>point</code>.
     */
    public double distanceTo(Point point) {
        return Math.sqrt(distanceSquaredTo(point));
    }
}
